extensions/cloudlink.js

```diff
@@ -1,1838 +1,2349 @@
-
-      // Rooms stuff
-      this.enableRoom = false;
-      this.isRoomSetting = false;
-      this.selectRoom = "";
-
-      // Remapping stuff
-      this.menuRemap = {
-        "Global data": "gmsg",
-        "Private data": "pmsg",
-        "Global variables": "gvar",
-        "Private variables": "pvar",
-        "Direct data": "direct",
-        "Status code": "statuscode",
-        "All data": "all",
-      };
+    // Establish a connection to the server
+    console.log("[CloudLink] Connecting to server:", url);
+    try {
+      clVars.socket = new WebSocket(url);
+    } catch (e) {
+      console.warn("[CloudLink] An exception has occurred:", e);
+      return;
+    }
+
+    // Bind connection established event
+    clVars.socket.onopen = function (event) {
+      clVars.currentServerUrl = url;
+      
+      // Set the link state to connected.
+      console.log("[CloudLink] Connected.");
+
+      clVars.linkState.status = 2;
+
+      // If a server_version message hasn't been received in over half a second, try to broadcast a handshake
+      clVars.handshakeTimeout = window.setTimeout(function() {
+        console.log("[CloudLink] Hmm... This server hasn't sent us it's server info. Going to attempt a handshake.");
+        sendHandshake();
+      }, 500);
+
+      // Fire event hats (only one not broken)
+      runtime.startHats('cloudlink_onConnect');
+
+      // Return promise (during setup)
+      return;
+    };
+
+    // Bind message handler event
+    clVars.socket.onmessage = function (event) {
+      handleMessage(event.data);
+    };
+
+    // Bind connection closed event
+    clVars.socket.onclose = function (event) {
+      switch (clVars.linkState.status) {
+        case 1: // Was connecting
+          // Set the link state to ungraceful disconnect.
+          console.log(`[CloudLink] Connection failed (${event.code}).`);
+          clVars.linkState.status = 4;
+          clVars.linkState.disconnectType = 1;
+          break;
+
+        case 2: // Was already connected
+          if (event.wasClean || clVars.linkState.isAttemptingGracefulDisconnect) {
+            // Set the link state to graceful disconnect.
+            console.log(`[CloudLink] Disconnected (${event.code} ${event.reason}).`);
+            clVars.linkState.status = 3;
+            clVars.linkState.disconnectType = 0;
+          } else {
+            // Set the link state to ungraceful disconnect.
+            console.log(`[CloudLink] Lost connection (${event.code} ${event.reason}).`);
+            clVars.linkState.status = 4;
+            clVars.linkState.disconnectType = 2;
+          }
+          break;
+      }
+
+      // Reset clVars values
+      resetOnClose();
+
+      // Run all onClose event blocks
+      runtime.startHats('cloudlink_onClose');
+      // Return promise (during setup)
+      return;
     }
+  }
+
+  // GET the serverList
+  try {
+    Scratch.fetch(
+      "https://mikedev101.github.io/cloudlink/serverlist.json"
```

Comment:
sorry for delay here, looks fine to me, just so you know, you can bypass the fetching permission prompt if you use `https://raw.githubusercontent.com/MikeDev101/cloudlink/master/serverlist.json` here instead

Reply:
We did that in the old version before this PR

I'd like to see that resolved before I merge it (I'll do it this week if you don't), otherwise everyone using this will now be getting a permission prompt before the project even starts, that's just not ideal

Reply:
Fixed in https://github.com/MikeDev101/cloudlink/commit/a5bcfa3cc58b8cd613c3babf9ab0703794a1b89c